l2-contracts/contracts/AttesterRegistry.sol

```diff
@@ -0,0 +1,159 @@
+        _;
+    }
+
+    constructor(address _owner) {
+        owner = _owner;
+    }
+
+    // Adds a new attester to the registry. Fails if an attester with the
+    // same public key already exists.
+    function add(
+        address nodeOwner,
+        uint256 weight,
+        bytes calldata pubKey
+    ) external onlyOwner {
+        // Check if an attester with the same node owner or public key already exists.
+        for (uint256 i = 0; i < attesterOwners.length; i++) {
+            require(attesterOwners[i] != nodeOwner, "nodeOwner already exists");
+            require(!compareBytes(attesters[attesterOwners[i]].pubKey, pubKey), "pubKey already exists");
+        }
+
+        attesters[nodeOwner] = Attester(weight, pubKey, false, 0);
+        attesterOwners.push(nodeOwner);
+    }
+
+    // Removes an attester. Should fail if the validator is still active or
+    // the inactivity delay has not passed.
+    function remove(address nodeOwner) external onlyOwner {
+        verifyExists(nodeOwner);
+        require(attesters[nodeOwner].isInactive, "Attester is still active");
+        require(epoch >= attesters[nodeOwner].inactiveSince + INACTIVITY_DELAY,
+            "Attester's inactivity delay has not passed"
+        );
+
+        // Remove from mapping.
+        delete attesters[nodeOwner];
+
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        attesterOwners[nodeOwnerIndex(nodeOwner)] = attesterOwners[attesterOwners.length - 1];
+        attesterOwners.pop();
+    }
+
+    // Inactivates an attester.
+    function inactivate(address nodeOwner) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].isInactive = true;
+        attesters[nodeOwner].inactiveSince = epoch;
+    }
+
+    // Activates an attester.
+    function activate(address nodeOwner) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].isInactive = false;
+        attesters[nodeOwner].inactiveSince = 0;
+    }
+
+    // Changes the weight.
+    function changeWeight(address nodeOwner, uint256 weight) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].weight = weight;
+    }
+
+    // Changes the public key.
+    function changePublicKey(address nodeOwner, bytes calldata pubKey) external onlyOwner {
+        verifyExists(nodeOwner);
+        attesters[nodeOwner].pubKey = pubKey;
+    }
+
+    // Creates a new committee list which will become the next committee list.
+    function setNextCommittee() external onlyOwner {
+        epoch += 1;
+
+        // Replace current committee with next committee.
+        delete committee;
+        for (uint256 i = 0; i < nextCommittee.length; i++) {
+            committee.push(nextCommittee[i]);
+        }
+
+        // Populate `nextCommittee` based on active attesters.
+        delete nextCommittee;
+        for (uint256 i = 0; i < attesterOwners.length; i++) {
```

Comment:
cache length

Reply:
change `i++` to `++i` to save gas

Reply:
Changes here: https://github.com/matter-labs/era-contracts/pull/555/commits/b86eaee9275e487f8690d2d14c9a1b23a12a50ba